Release notes for 2021-11-23 (#6983)
keystonejs/keystone: docs/pages/releases/2021-11-23.mdx

import { Markdown, getStaticProps } from '../../components/Markdown';
import { Emoji } from '../../components/primitives/Emoji';
import { Stack } from '../../components/primitives/Stack';

# Release: 23rd November 2021

The **Release Candidate** for Keystone 6 **General Availability** has arrived! Within you’ll find numerous improvements across the project. <Emoji symbol="⭐️" alt="Gold star"/>

> From this release, we’ll soon move **Keystone 6** to General Availability and graduate to the `@keystone-6` namespace on npm.

Reviewer: ```suggestion
After this release, we will be moving **Keystone 6** to General Availability and graduating to the `@keystone-6` namespace on npm.

```

**We highly recommend you upgrade to this release:**

```json
"keystone-next/keystone": "xx.x.x",
```

**⚠️ &nbsp; This release contains breaking changes! Please see below, and backup your data before upgrading.**

## Shorter Relationship Names <Emoji symbol="🤏" alt="Pinching hand" />

x> **⚠️ &nbsp; Breaking change! Please follow the guidance below to avoid losing data.**

The names of **one-sided** and **two-sided**, **many-many** relationships has been **shortened**. Two-sided many-many relationship names contain only the left-hand side names now; and the `_many` suffix has been dropped from one-sided many-many relationships.

This reduces the probability that you will exceed [PostgreSQL's 63 character limit for identifiers](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) with typical usage.

> There are two ways to update:

Reviewer: ```suggestion
There are two different methods of upgrading to this new version without losing your data
```

> - Set `db.relationName` on many to many relations, this allows your database to remain unchanged.

Reviewer: ```suggestion
- Explicitly set the `db.relationName` on many-to-many relations, allowing your database to remain unchanged.
```

import { Markdown, getStaticProps } from '../../components/Markdown';
import { Emoji } from '../../components/primitives/Emoji';
import { Stack } from '../../components/primitives/Stack';

# Release: 23rd November 2021

> Our **Release Candidate** for **General Availability** is finally here! This is our RC for **Keystone 6** with numerous tweaks and fixes across the project. <Emoji symbol="⭐️" alt="Gold star"/>

Reviewer 2: ```suggestion
The **Release Candidate** for Keystone 6 **General Availability** has arrived! Within you’ll find numerous improvements across the project. <Emoji symbol="⭐️" alt="Gold star"/>
```

> Next week we're planning to move **Keystone 6** to General Availability with this release, graduating to the `@keystone-6` namespace on npm.

Reviewer 2: ```suggestion
From this release, we’ll soon move **Keystone 6** to General Availability and graduate to the `@keystone-6` namespace on npm.
```

> We encourage all Keystone developers to upgrade to this release.

Reviewer 2: ```suggestion
**We highly recommend you upgrade to this release:**
```

```json
"keystone-next/keystone": "xx.x.x",
```

> **⚠️ &nbsp; This release contains breaking changes, please see below!**

Reviewer 2: ```suggestion
**⚠️ &nbsp; This release contains breaking changes! Please see below, and backup your data before upgrading.**
```

> !> Like this release? [Give us a star](https://github.com/keystonejs/keystone) on GitHub!

Reviewer 2: ```suggestion
```

## Shorter Relationship Names <Emoji symbol="🤏" alt="Pinching hand" />

> x> **⚠️ &nbsp; Breaking change, you may lose data if you do not follow the guide below.**

Reviewer 2: ```suggestion
x> **⚠️ &nbsp; Breaking change! Please follow the guidance below to avoid losing data.**
```

The names of **one-sided** and **two-sided**, **many-many** relationships has been **shortened**. Two-sided many-many relationship names contain only the left-hand side names now; and the `_many` suffix has been dropped from one-sided many-many relationships.

This reduces the probability that you will exceed [PostgreSQL's 63 character limit for identifiers](https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) with typical usage.

> There are two different methods of upgrading to this new version without losing your data

Reviewer: ```suggestion
There are two different ways you can update your schema:
```

- Explicitly set the `db.relationName` on many-to-many relations, allowing your database to remain unchanged.

> - Rename your many relation tables using a migration, updating your DB.

Reviewer: ```suggestion
- Rename your many relation tables using a migration, **changing your database**.
```

### Set `db.relationName` on many to many relations

Rather than doing a migration, you can set the new field property `db.relationName`, for either side of a many-to-many relationship field.

If set to the existing relation name, your database will remain unchanged.

For example, given a schema like this:

```ts
Post: list({
  fields: {
    tags: relationship({ ref: 'Tag.posts', many: true }),
  },
}),
Tag: list({
  fields: {
    posts: relationship({ ref: 'Post.tags', many: true }),
  },
}),
```

> Before this release, the generated Prisma schema looked like this:
> 
> ```prisma
> // This file is automatically generated by Keystone, do not modify it manually.
> // Modify your Keystone config when you want to change this.
> 
> datasource postgresql {
>   url      = env("DATABASE_URL")
>   provider = "postgresql"
> }
> 
> generator client {
>   provider   = "prisma-client-js"
>   output     = "node_modules/.prisma/client"
>   engineType = "binary"
> }
> 
> model Post {
>   id   String @id @default(cuid())
>   tags Tag[]  @relation("Post_tags_Tag_posts")
> }
> 
> model Tag {
>   id    String @id @default(cuid())
>   posts Post[] @relation("Post_tags_Tag_posts")
> }
> ```

Reviewer: ```suggestion
```

```

By adding `db: { relationName: 'Post_tags_Tag_posts' }` to one side of the many-to-many relationship; you can preclude yourself from a migration.

> **Note:** It doesn't matter which side of the relationship you put this property, but it should be only on one side; otherwise you will receive an error.

Reviewer 2: ```suggestion
?> It doesn't matter which side of the relationship you put this property, but it should be only on one side; otherwise you will receive an error.
```

```ts
Post: list({
  fields: {
    tags: relationship({ ref: 'Tag.posts', many: true, db: { relationName: 'Post_tags_Tag_posts' } }),
  },
}),
Tag: list({
  fields: {
    posts: relationship({ ref: 'Post.tags', many: true }),
  },
}),
```

### Rename your many relation tables using a migration

For example, given a schema like this:

```ts
Post: list({
  fields: {
    tags: relationship({ ref: 'Tag.posts', many: true }),
  },
}),
Tag: list({
  fields: {
    posts: relationship({ ref: 'Post.tags', many: true }),
  },
}),
```

When updating to this change, and running `yarn dev`, Keystone will prompt you to update your schema.

x> <Emoji symbol="⚠️" alt="Danger" /> **Warning: DO NOT APPLY THE AUTOMATICALLY GENERATED MIGRATION!**<br/> You will lose your data. Only apply the migration if you want to `DROP` your data.

> If using `useMigrations: true`, Keystone will follow the typical migration flow offer to apply an automatically generated migration.

Reviewer 2: ```suggestion
If using `useMigrations: true`, Keystone will follow the typical migration flow and offer to apply an automatically generated migration.
```

-- CreateIndex
CREATE INDEX "_Post_tags_B_index" ON "_Post_tags"("B");
```

You need to modify it so that it looks like this with the old and new table names for your schema substituted:

```sql
ALTER TABLE "_Post_tags_Tag_posts" RENAME TO "_Post_tags";
DROP INDEX "_Post_tags_Tag_posts_AB_unique";
DROP INDEX "_Post_tags_Tag_posts_B_index";
CREATE UNIQUE INDEX "_Post_tags_AB_unique" ON "_Post_tags"("A", "B");
CREATE INDEX "_Post_tags_B_index" ON "_Post_tags"("B");
```

## Query Engine Switch <Emoji symbol="🚂" alt="Locomotive engine" />

Keystone now uses Prisma's Node-API Query Engine instead of the Binary Query Engine. This should improve the performance of operations using Prisma.

From our initial testing, performance has increased significantly when getting large amounts of data and is marginally better for smaller amounts.

See the [Prisma docs](https://www.prisma.io/docs/concepts/components/prisma-engines/query-engine) for more details.

## Node Engines <Emoji symbol="💽" alt="Computer disc" />

Keystone officially supports running on LTS versions of Node.js this is currently version 14 and 16. We've updated our engine values throughout the project to reflect this.

> We recommend you run Node.js `^14.15` or `^16.13` for the best Keystone experience.

Reviewer 2: ```suggestion
!> We recommend you run Node.js `^14.15` or `^16.13` for the best Keystone experience.
```

- Fixed clear button not removing inline relationships when used in the document field.

- Relationship field now respects `ui.displayMode: 'cards'` in the create view.

- The `Set as Authenticated Item`/`Add Authenticated Item` button is now hidden if the relationship field already has the authenticated item.

- Fixed `ui.isAccessAllowed` not being respected in the admin meta query when no session strategy was defined.

- The item page in the Admin UI no longer crashes when failing to fetch an item.

- The admin meta query now bypasses `ui.isAccessAllowed` for sudo contexts.

- Fixed doing multiple writes at the same time on SQLite causing an timeout immediately.

## Credits <Emoji symbol="💫" alt="Circle Star" />

- Thanks to [gregpalaci](https://github.com/gregpalaci), [vicaia](https://github.com/vicaia), [jakegiri](https://github.com/vicaia) [davebeauchemin](https://github.com/davebeauchemin), [Temkit](https://github.com/Temkit), and [bkbooth](https://github.com/bkbooth) for reporting bugs that were fixed in this release!

- Fixed default renderer of `layout` block in `<DocumentRenderer />`. Thanks [@datner](https://github.com/datner)!

> 

Reviewer 2: ```suggestion
!> Like this release? [Give us a star](https://github.com/keystonejs/keystone) on GitHub!
```
